File: VB/CRUDBehavior/XPOInstantModeCRUDBehavior.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Windows.Interactivity
Imports System.Windows
Imports DevExpress.Xpf.Grid
Imports System.Windows.Controls
Imports System.Windows.Input
Imports DevExpress.Xpf.Core
Imports System.ComponentModel
Imports DevExpress.Xpo
Imports CRUDBehaviorBase

Namespace XPOInstant
    Public Class XPOInstantModeCRUDBehavior
        Inherits CRUDBehaviorBase.CRUDBehaviorBase

        Public Shared ReadOnly InstantCollectionSourceProperty As DependencyProperty = DependencyProperty.Register("InstantCollectionSource", GetType(XPInstantFeedbackSource), GetType(XPOInstantModeCRUDBehavior), New PropertyMetadata(Nothing))
        Public Property InstantCollectionSource() As XPInstantFeedbackSource
            Get
                Return CType(GetValue(InstantCollectionSourceProperty), XPInstantFeedbackSource)
            End Get
            Set(ByVal value As XPInstantFeedbackSource)
                SetValue(InstantCollectionSourceProperty, value)
            End Set
        End Property
        Protected Overrides Function CanExecuteRemoveRowCommand() As Boolean
            If CollectionSource Is Nothing OrElse Grid Is Nothing OrElse View Is Nothing OrElse Grid.CurrentItem Is Nothing Then
                Return False
            End If
            Return True
        End Function
        Protected Overrides Sub Initialize()
            Grid.ItemsSource = InstantCollectionSource
            NewRowCommand.RaiseCanExecuteChangedEvent()
            CollectionSource = New XPServerCollectionSource(New UnitOfWork(), InstantCollectionSource.ObjectType) With {.AllowNew = True, .AllowRemove = True, .AllowEdit = True, .TrackChanges = True, .DeleteObjectOnRemove = True}
            MyBase.Initialize()
        End Sub
        Protected Overrides Sub UpdateDataSource()
            InstantCollectionSource.Refresh()
        End Sub
        Protected Overrides Sub OnAttached()
            MyBase.OnAttached()
            If View IsNot Nothing AndAlso InstantCollectionSource IsNot Nothing Then
                Initialize()
            Else
                AddHandler Grid.Loaded, AddressOf OnGridLoaded
            End If
        End Sub
    End Class
End Namespace

File: CS/CRUDBehavior/XPOInstantModeCRUDBehavior.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows.Interactivity;
using System.Windows;
using DevExpress.Xpf.Grid;
using System.Windows.Controls;
using System.Windows.Input;
using DevExpress.Xpf.Core;
using System.ComponentModel;
using DevExpress.Xpo;
using CRUDBehaviorBase;

namespace XPOInstant {
    public class XPOInstantModeCRUDBehavior: CRUDBehaviorBase.CRUDBehaviorBase {
        public static readonly DependencyProperty InstantCollectionSourceProperty =
            DependencyProperty.Register("InstantCollectionSource", typeof(XPInstantFeedbackSource), typeof(XPOInstantModeCRUDBehavior), new PropertyMetadata(null));
        public XPInstantFeedbackSource InstantCollectionSource {
            get { return (XPInstantFeedbackSource)GetValue(InstantCollectionSourceProperty); }
            set { SetValue(InstantCollectionSourceProperty, value); }
        }
        protected override bool CanExecuteRemoveRowCommand() {
            if(CollectionSource == null || Grid == null || View == null || Grid.CurrentItem == null) return false;
            return true;
        }
        protected override void Initialize() {
            Grid.ItemsSource = InstantCollectionSource;
            NewRowCommand.RaiseCanExecuteChangedEvent();
            CollectionSource = new XPServerCollectionSource(new UnitOfWork(), InstantCollectionSource.ObjectType) {
                AllowNew = true,
                AllowRemove = true,
                AllowEdit = true,
                TrackChanges = true,
                DeleteObjectOnRemove = true
            };
            base.Initialize();
        }
        protected override void UpdateDataSource() {
            InstantCollectionSource.Refresh();
        }
        protected override void OnAttached() {
            base.OnAttached();
            if(View != null && InstantCollectionSource != null)
                Initialize();
            else
                Grid.Loaded += OnGridLoaded;
        }
    }
}